runner/local/queueing_runner.go

```diff
@@ -0,0 +1,228 @@
+	nextRunId         int // used to assign unique runIds to the run requests
+	runQueue          []commandAndId
+	runnerIsAvailable bool
+
+	notifyRunnerAvailCh chan struct{} // the runner will use this channel to signal it is available
+
+	runRequestsCh chan runRequestEvent // channel for synchronizing calls to Run()
+
+	statusRequestsCh chan string          // channel for synchronizing status requests
+	statusResponseCh chan requestResponse // channel for unblocking the status requests
+
+	// map the queueing runner run ids to the runner's run ids. We need runner's run id for Status()
+	queueIdToRunnerIdMap map[string]string
+}
+
+// This must be used to initialize QueueingRunner properly
+func NewQueuingRunner(context context.Context,
+	runner runner.Runner,
+	maxQueueLen int,
+	runnerAvailableCh chan struct{}) runner.Runner {
+
+	qRunner := &QueueingRunner{
+		runner:               runner,
+		done:                 false,
+		maxQueueLen:          maxQueueLen,
+		nextRunId:            0,
+		runnerIsAvailable:    true,
+		notifyRunnerAvailCh:  runnerAvailableCh,
+		runRequestsCh:        make(chan runRequestEvent),
+		statusRequestsCh:     make(chan string),
+		statusResponseCh:     make(chan requestResponse),
+		queueIdToRunnerIdMap: make(map[string]string),
+	}
+
+	// start the request event processor
+	go qRunner.eventLoop()
+
+	return qRunner
+}
+
+// Put the request and a callback channel (created for the request)
+// on the queueing runner's processRequests channel.  When the request has been
+// queued the request's runid, its queued status and any error will be returned
+// via the request's callback channel
+func (qr *QueueingRunner) Run(c *runner.Command) (runner.ProcessStatus, error) {
+	log.Printf("local.queueingRunner: in qr.Run() args:%v\n", c.Argv)
+
+	// make a channel for the 'on queue' response
+	requestOnQueueCh := make(chan requestResponse)
+
+	// put the request on the process requests channel
+	requestContent := runRequestEvent{cmd: c, onQueueResponseCh: requestOnQueueCh}
+	qr.runRequestsCh <- requestContent
+
+	// get and return the 'on queue' response from this request's channel
+	requestQueued := <-requestOnQueueCh
+	close(requestOnQueueCh)
+	return *requestQueued.status, requestQueued.err
+}
+
+// The 'events' include: Run() request, Status() request, or a signal that the runner is available.
+// Run() events are put on the queue, Status() and runner available events are processed immediately
+// Cycles that don't process Run(), Status() or runner avaialable events (default case) are used to
+// run the next command in the queue.
+func (qr *QueueingRunner) eventLoop() {
+	for !qr.done {
+
+		select {
+		case nextRequest := <-qr.runRequestsCh:
+			qr.addRequestToQueue(nextRequest)
+
+		case <-qr.notifyRunnerAvailCh:
+			qr.runnerIsAvailable = true
+
+		case qRunId := <-qr.statusRequestsCh:
+			runnerRunId := qr.queueIdToRunnerIdMap[qRunId]
+			qr.getRunStatus(runner.RunId(runnerRunId), runner.RunId(qRunId))
+
+		default:
```

Comment:
This seems like it'll peg the cpu. We likely don't want a default case and can just move qr.RunNextCommandInQueue directly into case <-qr.notifyRunnerAvailCh.